bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java

```diff
@@ -0,0 +1,433 @@
+                if (jo.has("url")) {
+                    return jo.get("url").getAsString();
+                } else {
+                    errorMsg = "Could not get url";
+                }
+            } else {
+                errorMsg = "Could not get result";
+            }
+        } else {
+            errorMsg = "Received message is invalid JSON";
+        }
+        logger.debug("{}: {}", errorMsg, mapResponse);
+        return "";
+    }
+
+    public String getDeviceStatus(String device, String country) throws MiCloudException {
+        String url = getApiUrl(country) + "/home/device_list";
+        Map<String, String> map = new HashMap<String, String>();
+        map.put("data", "{\"dids\":[\"" + device + "\"]}");
+        final String response = request(url, map);
+        logger.debug("response: {}", response);
+        return response;
+    }
+
+    public String getDevices(String country) {
+        String url = getApiUrl(country) + "/home/device_list";
+        Map<String, String> map = new HashMap<String, String>();
+        map.put("data", "{\"getVirtualModel\":false,\"getHuamiDevices\":0}");
+        String resp;
+        try {
+            resp = request(url, map);
+            logger.trace("Get devices response: {}", resp);
+            CloudUtil.printDevices(resp, country, logger);
+            if (resp.length() > 2) {
+                CloudUtil.saveFile(resp, country, logger);
+                return resp;
+            }
+        } catch (MiCloudException e) {
+            logger.info("{}", e.getMessage());
+        }
+        return "";
+    }
+
+    public String request(String urlPart, String country, Map<String, String> params) throws MiCloudException {
+        String url = getApiUrl(country) + urlPart;
+        String response = request(url, params);
+        logger.debug("Request to {} server {}. Response: {}", country, urlPart, response);
+        return response;
+    }
+
+    public String request(String url, Map<String, String> params) throws MiCloudException {
+        if (this.serviceToken.isEmpty() || this.userId.isEmpty()) {
+            throw new MiCloudException("Cannot execute request. service token or userId missing");
+        }
+        startClient();
+        logger.debug("Send request: {} to {}", params.get("data"), url);
+        Request request = httpClient.newRequest(url).timeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        request.agent(USERAGENT);
+        request.header("x-xiaomi-protocal-flag-cli", "PROTOCAL-HTTP2");
+        request.header(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded");
+        request.cookie(new HttpCookie("userId", this.userId));
+        request.cookie(new HttpCookie("yetAnotherServiceToken", this.serviceToken));
+        request.cookie(new HttpCookie("serviceToken", this.serviceToken));
+        request.cookie(new HttpCookie("locale", locale.toString()));
+        request.cookie(new HttpCookie("timezone", "GMT%2B01%3A00"));
```

Comment:
Shouldn't the timezone be the local timezone, or no?

Reply:
Probable indeed... this stuff is mostly from reverse engineering and MITM logging of traffic from phone to the cloud. It is unclear what it actually is doing. I'll make it dynamic based on the OS setting